8332313: Update code review guidelines

kevinrushforth opened the pull request:
Update the code review guidelines for JavaFX.

The JavaFX [CONTRIBUTING](https://github.com/kevinrushforth/jfx/blob/8332313-contributing/CONTRIBUTING.md) guidelines includes guidance for creating, reviewing, and integrating changes to JavaFX, along with a pointer to a [Code Review Policies](https://wiki.openjdk.org/display/OpenJFX/Code+Reviews) Wiki page.

This PR updates these guidelines to improve the quality of reviews, with a goal of improving JavaFX and decreasing the chance of introducing a serious regression or other critical bug.

The source branch has three commits:
1. Converts the Code Review Policies Wiki page to a [README-code-reviews.md](https://github.com/kevinrushforth/jfx/blob/8332313-contributing/README-code-reviews.md) file in the repo and updates hyperlinks to the new location.
2. Update `README-code-reviews.md` with new guidelines
3. Update `CONTRIBUTING.md` to highlight important requirements  (and minor changes to `README-code-reviews.md`)

Commit 1 is content neutral, so it might be helpful for reviewers to look at the changes starting from the second commit.

The updates are:

* In the Overview section, add a list of items for Reviewers, PR authors, and sponsoring Committers to verify prior to integration
* Create a "Guidelines for reviewing a PR" subsection of the Code review policies section
* Create a "Before you integrate or sponsor a PR" subsection of the Code review policies section
* Update the `CONTRIBUTING.md` page to highlight important requirements

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (3 reviews required, with at least 3 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8332313](https://bugs.openjdk.org/browse/JDK-8332313): Update code review guidelines (**Enhancement** - P3)


### Reviewers
 * [John Hendrikx](https://openjdk.org/census#jhendrikx) (@hjohn - Committer)
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Nir Lisker](https://openjdk.org/census#nlisker) (@nlisker - **Reviewer**)
 * [Johan Vos](https://openjdk.org/census#jvos) (@johanvos - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1455/head:pull/1455` \
`$ git checkout pull/1455`

Update a local copy of the PR: \
`$ git checkout pull/1455` \
`$ git pull https://git.openjdk.org/jfx.git pull/1455/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1455`

View PR using the GUI difftool: \
`$ git pr show -t 1455`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1455.diff">https://git.openjdk.org/jfx/pull/1455.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1455#issuecomment-2113232915)

kevinrushforth commented:
Reviewers: @johanvos @andy-goryachev-oracle @arapte @prrace 

I'd like at least 3 "R"eviewers to review this. Committers and Authors are also welcome to provide feedback.

/reviewers 3 reviewers

andy-goryachev-oracle commented on CONTRIBUTING.md:
```diff
@@ -46,7 +46,7 @@ If you are a first time contributor to OpenJFX, welcome! Please do the following
+    Please read the entire section below on how to submit a pull request, as well as the [Code Review Policies](README-code-reviews.md) page. If this is a feature request, please note the additional requirements and expectations in the [New features / API additions](#new-features--api-additions) section at the end of this guide.
 
 * File a bug in JBS for every pull request
 
```
for line 53, would it make more sense to use this link
https://bugs.openjdk.org/projects/JDK/issues/
?

kevinrushforth replied:
Whether or not that might be a better link, it's unrelated to this PR. The scope of this PR is limited to guidelines surrounding reviewing, integrating, and sponsoring PRs.

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+The OpenJFX Project is guided by the Project Leads and Reviewers.
+
+__Project Co-Lead__: Kevin Rushforth (kcr) <br>
+__Project Co-Lead__: Johan Vos (jvos)
```
There are two sets of ids - one for OpenJFX/JBS and one for Github.  This might be confusing sometimes, should we list both?

kevinrushforth replied:
Not sure what you mean, but it doesn't seem related to this PR.

prrace replied:
He means people reading reviews are usually on github and your github ID is kevinrusforth,
so they might be confused by "kcr" which is your openjdk id.

Since the context is identifying OpenJDK Project leads, I think the OpenJDK ID is the right and only one to include here. At most you could clarify  like (OpenJDK ID - kcr)

andy-goryachev-oracle replied:
exactly my point.  when the majority of time is spent in the context of github (PR review etc) it might be confusing when one types `@jvos` and it's not found, or when within a PR there are references to both `jvos` and `johanvos`

(see the very first comment in https://github.com/openjdk/jfx/pull/1388 )

kevinrushforth replied:
fixed

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+Here is a list of things to keep in mind when reviewing a PR. This applies to anyone doing a review, but especially a "R"eviewer:
+
+* Make sure you understand why there was an issue to begin with, and why/how the proposed PR solves the issue
```
Perhaps this should be noted as a required step for the PR author (maybe as a template text in the PR description?).

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+Here is a list of things to keep in mind when reviewing a PR. This applies to anyone doing a review, but especially a "R"eviewer:
+
+* Make sure you understand why there was an issue to begin with, and why/how the proposed PR solves the issue
+* Focus first on substantive comments rather than stylistic comments
```
this is not as important as the next two.

kevinrushforth replied:
fixed

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+* Make sure you understand why there was an issue to begin with, and why/how the proposed PR solves the issue
+* Focus first on substantive comments rather than stylistic comments
+* Consider the risk of regression
+* Consider any compatibility concerns
```
regression and compatibility risks are probably the most important aspects of the code review from the "R"eviewer's perspective.  Perhaps this should be emphasized somehow?

johanvos replied:
In the ideal world where we have tons of regression and compatibility tests, I would agree. Unfortunately, we are totally not there yet. Compared to other projects, the quality of tests in OpenJFX is good, but given the multitude of usage scenario's in client development, we would need much, much more tests before we can rely on them to detect regression.
Therefore, the #1 point imho is that both the author (as Andy commented) and the reviewer have a real good understanding on what is happening. Every single change in a PR should be understood. The most dangerous things are "I don't understand why, but it seems to fail before and succeed after the patch".

kevinrushforth replied:
Yes, this a good point. I left the "make sure you understand..." bullet as the first, and moved the "Focus first on substantive comments" below the regression and compatibility bullets.

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+* All substantive feedback has been addressed, especially any objections from one with a Reviewer role.
+* All Reviewers who have requested the chance to review have done so (or indicated that they are OK with it going in without their review). In rare cases a Project Lead may override this.
+* The PR has been "rfr" (as indicated by Skara) for at least 1 business day (24 hours), not including weekends / or major holidays. This is to allow sufficient time for those reviewers who might be in other time zones the chance to review if they have concerns. This is measured from the time that Skara has most recently added the "rfr" label (for example, for a PR that was previously in Draft mod ...
```
nit pick: 24 hours does not always equal to one business day.  A long weekend may extend the waiting period to 96 hours, a "winter break" might span the whole week.

kevinrushforth replied:
fixed

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+* All substantive feedback has been addressed, especially any objections from one with a Reviewer role.
+* All Reviewers who have requested the chance to review have done so (or indicated that they are OK with it going in without their review). In rare cases a Project Lead may override this.
+* The PR has been "rfr" (as indicated by Skara) for at least 1 business day (24 hours), not including weekends / or major holidays. This is to allow sufficient time for those reviewers who might be in other time zones the chance to review if they have concerns. This is measured from the time that Skara has most recently added the "rfr" label (for example, for a PR that was previously in Draft mod ...
+* Double-check the commit message before you integrate. Skara will list it near the top of your PR.
```
is this needed?  SKARA makes sure that it matches the JBS title.

kevinrushforth replied:
removed bullet

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+#### A. Low-impact bug fixes.
+
+These are typically isolated bug fixes with little or no impact beyond fixing the bug in question; included in this category are test fixes (including most new tests), doc fixes, and fixes to sample applications (including most new samples).
```
I think the main criteria is regression and compatibility impact.  Should we expand upon that here?

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+A feature should be discussed up-front on the openjfx-dev mailing list to get early feedback on the concept (is it a feature we are likely to accept) and the direction the API and/or implementation should take.
+
+To ensure that new features are consistent with the rest of the API and the desired direction of the Project, a CSR is required for a new Feature, API addition, or behavioral change. The CSR must be reviewed and approved by a "lead" of the Project. Currently this is either Kevin Rushforth or Johan Vos as indicated above.
```
this paragraph feels extremely nebulous.

a minor issue is that CSR cannot be written in full detail before the draft PR is created (a PoS is written),

a larger issue is that a decision to implement a feature might need a better process - is one lead approval sufficient?  what happens when leads disagree?

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+Feature requests come with a responsibility beyond just saying "here is the code for this cool new feature, please take it". There are many factors to consider for even small features. Larger features will need a significant contribution in terms of API design, coding, testing, maintainability, etc.
+
+A feature should be discussed up-front on the openjfx-dev mailing list to get early feedback on the concept (is it a feature we are likely to accept) and the direction the API and/or implementation should take.
```
are there any special rules to gauging the consensus?  what happens when one party in a discussion stops responding?

andy-goryachev-oracle commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+To ensure that new features are consistent with the rest of the API and the desired direction of the Project, a CSR is required for a new Feature, API addition, or behavioral change. The CSR must be reviewed and approved by a "lead" of the Project. Currently this is either Kevin Rushforth or Johan Vos as indicated above.
+
+The review of the implementation follows the same "two reviewer" standard for higher-impact changes as described in category B. The two code reviewers for the implementation may or may not include the Lead who reviewed the CSR. The review / approval of the CSR is an independent step from the review / approval of the code change, although they can proceed in parallel.
```
it is not clear (at least to me) the required order in respect to CSR approval.  for a new feature, does the CSR need to be approved before the work can be started, or the draft PR is published, or an PR enters its "rfr" stage, or just before the integration?

Or perhaps it should be clarified that the CSR is just a human-readable, specially formatted specification of the *change*, and it is an integral part of the new feature that needs to be approved before the feature can be integrated?

kevinrushforth replied:
This is out of scope for this PR, but it might be something to consider addressing in a follow-on enhancement.

prrace replied:
It should be made clear (somewhere, at some point) that it would be rare for a CSR to be completed and approved before starting work on the implementation because the spec. almost always evolves as the fix is developed.

kevinrushforth replied:
Yes, that is a good point. I'll file a follow-up to do that (since this section is otherwise unmodified from the current guidelines on the Wiki).

nlisker commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+All code reviews must be done via a pull request submitted against this GitHub repo, [openjdk/jfx](https://github.com/openjdk/jfx). A JBS bug ID must exist before the pull request will be reviewed. See [CONTRIBUTING.md](CONTRIBUTING.md) for information on how to submit a pull request.
+
+All fixes must be reviewed by at least one reviewer with the "Reviewer" role (aka a "R"eviewer). We have a different code review threshold for different types of changes. If there is disagreement as to whether a fix is low-impact or high-impact, then it is considered high-impact. In other words we will always err on the side of quality by "rounding up" to the next higher category. The contributor ...
```
"but **It** is" -> it

prrace replied:
I think it worth noting that in skara syntax that isn't two people with the reviewer role.
And tell people what to use if that is what they intend - eg if I have it right
/reviewers reviewers 2

kevinrushforth replied:
It's `/reviewers 2 reviewers`, so I'll add that as an example and clarify that `/reviewers 2` is 2 with at least one reviewer.

kevinrushforth replied:
Fixed (both the example Phil asked for and the typo Nir noted)

nlisker commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+All code reviews must be done via a pull request submitted against this GitHub repo, [openjdk/jfx](https://github.com/openjdk/jfx). A JBS bug ID must exist before the pull request will be reviewed. See [CONTRIBUTING.md](CONTRIBUTING.md) for information on how to submit a pull request.
+
+All fixes must be reviewed by at least one reviewer with the "Reviewer" role (aka a "R"eviewer). We have a different code review threshold for different types of changes. If there is disagreement as to whether a fix is low-impact or high-impact, then it is considered high-impact. In other words we will always err on the side of quality by "rounding up" to the next higher category. The contributor ...
```
About requesting reviews. I think that only some people can request reviews through GitHub, I never managed to do it on this repo, probably a matter of permissions. Might worth clarifying how this works.

kevinrushforth replied:
It's better to just add comments `@` mentioning who you want to do the review, since there isn't a way for most people to request a review from others.

nlisker commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+* Consider the risk of regression
+* Consider any compatibility concerns
+* Check whether there is an automated test; if not, ask for one, if it is feasible
+* Make sure that the PR has executed the GHA tests and that they all pass; if they aren't being run, ask the PR author to enable GHA workflows
```
These tests sometimes fail and we integrate anyway. I noticed that sometimes they need a few iterations to succeed. Are we really dependent on them?

andy-goryachev-oracle replied:
or is it?  :-)

hjohn replied:
Isn't this automatic? Seems weird you could integrate without these passing.

nlisker replied:
> or is it? :-)

![image](https://github.com/openjdk/jfx/assets/37422899/8daab7cf-f050-4964-b8a6-731666422293)

Looks to me like it is...

kevinrushforth replied:
A passing GHA test run is neither necessary nor sufficient. It is an interesting data point. A PR can be integrated with a failing run as long as we understand why it failed. It is something for a reviewer to check and ask about if they suspect it points to a real problem. I'll reword this a little bit to make it clear. I'm sort of glad that GitHub / Azure  chose today to break it. It is a great illustration of why we don't want it to be blocking.

Btw, I filed https://bugs.openjdk.org/browse/JDK-8332328 to track the GHA failure.

kevinrushforth replied:
I updated the wording slightly

nlisker commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+* Consider any compatibility concerns
+* Check whether there is an automated test; if not, ask for one, if it is feasible
+* Make sure that the PR has executed the GHA tests and that they all pass; if they aren't being run, ask the PR author to enable GHA workflows
+* If the PR source branch hasn't synced up from master in a long time, or if there is an upstream commit not in the source branch that might interfere with the PR, ask the PR author to merge the latest upstream master.
```
This is the only bullet point with a period at the end.

kevinrushforth replied:
I'll fix it.

kevinrushforth replied:
fixed

nlisker commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+* If you want to indicate your approval, but still feel additional reviewers are needed, you may increase the number of reviewers (e.g., from 2 to 3)
+* If you want an area expert to review a PR, indicate this in a comment of the form: `Reviewers: @PERSON1 @PERSON2`; the requested reviewers can indicate whether or not they plan to review it
+* If you want to ensure that you have the opportunity to review this PR yourself, add a comment of the form: `@PRAUTHOR Wait for me to review this PR`, optionally add any concerns you might have
+
```
I would add that it's a good idea to search for JBS issues similar/linked to the one being targeted. There are many unfound duplicates in JBS that arise only when searching, or at least closely linked ones (problems in `TableTreeView` often have "pseudo-duplicates" for `TreeView` and `TableView`, for example). 

I would explain (or link to where it is explained) when to close an issue a a duplicated vs. when to use the /add Skara command to bundle issues into the same fix.

We might also find that the targeted issue is a regression of another issue. This is somewhat mentioned in the bullet point below "Consider the risk of regression".

andy-goryachev-oracle replied:
Second this, although I think the bulk of the work should fall on the author of the PR.

kevinrushforth replied:
This would probably be better in the CONTRIBUTING guidelines (for PR authors). Maybe as a follow-on, but I'll give it some thought.

nlisker commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+* If you want to indicate your approval, but still feel additional reviewers are needed, you may increase the number of reviewers (e.g., from 2 to 3)
+* If you want an area expert to review a PR, indicate this in a comment of the form: `Reviewers: @PERSON1 @PERSON2`; the requested reviewers can indicate whether or not they plan to review it
+* If you want to ensure that you have the opportunity to review this PR yourself, add a comment of the form: `@PRAUTHOR Wait for me to review this PR`, optionally add any concerns you might have
+
```
Another thing to look out for is the targeting branch during rampdown. Should they all target master and then be backported as needed, or can some target the rampdown branch?

kevinrushforth replied:
Good idea. I'll add the bit about checking the target branch -- if the PR isn't a backport it should almost always target the master branch (there might be _very_ rare exception where a bug is specific to the stabilization branch).

kevinrushforth replied:
fixed

nlisker commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+* Make sure you understand why there was an issue to begin with, and why/how the proposed PR solves the issue
+* Focus first on substantive comments rather than stylistic comments
+* Consider the risk of regression
+* Consider any compatibility concerns
```
This might be included in this point already, but one thing I sometimes miss is the inadvertent introduction of new API (that will have to be deprecated if missed). This can happen with `protected` methods, default `public` constructors (esp. if a non-API constructor is removed), or if a class is moved from a non-exported to an exported package (something that you can't see by looking at the area you're reviewing, you need to check the `module-info`, or more practically, look at the package names).

andy-goryachev-oracle replied:
I was wondering if there ought to be an unofficial checklist for things to look for?
As new people become "R"eviewers, I think there is a value in accumulating collective wisdom in a checklist.  I like checklists.

kevinrushforth replied:
> ...inadvertent introduction of new API (that will have to be deprecated if missed)

I think this is worth mentioning.

> I was wondering if there ought to be an unofficial checklist for things to look for?

That's sort of what this section is meant to be. As long as it doesn't get too verbose we could add additional things to look for.

kevinrushforth replied:
> > ...inadvertent introduction of new API (that will have to be deprecated if missed)
> 
> I think this is worth mentioning.

Fixed.

nlisker commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+* Determine whether this needs 2 reviewers and whether it needs a CSR; issue the `/reviewers 2` or `/csr` command as needed
+* If you want to indicate your approval, but still feel additional reviewers are needed, you may increase the number of reviewers (e.g., from 2 to 3)
+* If you want an area expert to review a PR, indicate this in a comment of the form: `Reviewers: @PERSON1 @PERSON2`; the requested reviewers can indicate whether or not they plan to review it
```
Should a list of experts per area be available somewhere? The Wiki has an old "code ownership" table that is out of date. Usually you get to know the experts only after they have reviewed your code a couple of times.

kevinrushforth replied:
That might be something to consider as a follow-on.

prrace commented on CONTRIBUTING.md:
```diff
@@ -229,6 +230,8 @@ Please also follow these formatting guidelines:
 * Wildcard imports &ndash; for example, `import java.util.*;` &ndash; are forbidden and may cause the build to fail. Please attempt to configure your IDE so it doesn't generate wildcard imports. An exception to this rule is that wildcard static imports in test classes are allowed, for example, `import static org.junit.Assert.*;`.
 * Don't worry too much about import order. Try not to change it but don't worry about fighting your IDE to stop it from doing so.
 
+New code should be formatted consistently in accordance with the above guidelines. However, please do not reformat existing code as part of a bug fix. The makes more changes for code reviewers to track and review, and can lead to merge conflicts. If you want to reformat a class, do that in a separate pull request (which will need its own unique JBS bug ID).
```
"The makes more changes" ? I think you mean "This" not "The"

I'm not sure about the last sentence, it seems to encourage reformatting fixes which are just noise most of the time.

johanvos replied:
I agree with the concern, but I still think it's much better to encourage developers to do formatting in a separate issue (or not at all) with all the required administration, than to sneak in a formatting change in a PR that has nothing to do with formatting. I prefer the noise to be completely isolated, so that it can be ignored easily, rather than being distracted by it in a PR.

hjohn replied:
I think this conflicts a bit with the previous statement ("don't worry too much about import order"). In my experience, when merging complex changes, import order changes are often the cause of conflicts. We also found that with many different committer preferences, the whole import block could change from one commit to the next, which will lead to a conflict if such a file is part of a backport.

We solved this by simply mandating a single order for imports (alphabetical, no gaps, no star imports) and made the build fail when it was violated.  The number of conflicts went down dramatically.  This does require a one time import fix of the code base, which will also be a source of conflicts, but at least it will be in the usual area (imports) and should solve the problem once and for all.

kevinrushforth replied:
I removed the bit encouraging reformatting PRs (but left the "don't reformat existing code..." bit.

As for the import order, this came up recently in Andy's PRs to remove unnecessary imports. I'll file a follow-on issue to evaluate whether we want to change our current policy.

hjohn commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+* All substantive feedback has been addressed, especially any objections from one with a Reviewer role.
+* All Reviewers who have requested the chance to review have done so (or indicated that they are OK with it going in without their review). In rare cases a Project Lead may override this.
+* The PR has been "rfr" (as indicated by Skara) for at least 1 business day (24 hours), not including weekends or major holidays. This is to allow sufficient time for those reviewers who might be in other time zones the chance to review if they have concerns. This is measured from the time that Skara has most recently added the "rfr" label (for example, for a PR that was previously in Draft mode, ...
+* Verify the commit message. The Skara tooling adds a comment near the top of the PR with the commit message that will be used. You can add a summary to the commit message with the `/summary` command. You can add additional contributors with the `/contributor` command. Commands are issued by adding a comment to the PR that starts with a slash `/` character.
```
The last sentence feels a bit out of place, as several places earlier in this document already mention slash commands.

kevinrushforth replied:
I'll remove it (I moved it from the CONTRIBUTING doc, and  it does seem out of place).

Actually as someone else commented, Skara will verify that the PR title and JBS bug match, so mostly this is about adding a summary and/or contributors if you want to.

kevinrushforth replied:
Removed.

prrace commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+### Reviewers
+
+The [List of Reviewers](https://openjdk.java.net/census#openjfx) is on the OpenJDK Census.
```
We use ".org" now, not ".java.net"

kevinrushforth replied:
fixed

prrace commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+
+### 1. The Reviewer role for the OpenJFX Project
+
+We define a formal "Reviewer" role, similar to the JDK project. A [Reviewer](https://openjdk.java.net/census#openjfx) is responsible for reviewing code changes and helping to determine whether a change is suitable for including into OpenJFX. We expect Reviewers to feel responsible not just for their piece, but for the quality of the JavaFX library as a whole. In other words, the role of Reviewer  ...
```
(https://openjdk.java.net/census#openjfx)

.org please

BTW these very long source lines make it awkward to precisely identify the text I'm commenting on.

kevinrushforth replied:
I'll fix it.

kevinrushforth replied:
fixed

prrace commented on README-code-reviews.md:
```diff
@@ -0,0 +1,103 @@
+Skara will mark a PR as "ready" once the minimum number of reviewers have reviewed it. Before you integrate or sponsor the PR, ensure that the following have been done:
+
+* All substantive feedback has been addressed, especially any objections from one with a Reviewer role.
+* All Reviewers who have requested the chance to review have done so (or indicated that they are OK with it going in without their review). In rare cases a Project Lead may override this.
```
One thing to add here (or hereabouts) is that if someone has commented on your review and requested changes that in almost all cases you should expect that they will want to return to review the results. So DO NOT push without letting earlier reviewers who made substantive comments re-review.

kevinrushforth replied:
I'll add something about this.

kevinrushforth replied:
fixed

kevinrushforth reviewed (commented):
Thanks for the additional feedback. I'll update the PR in the next day or two. I'll double-check that I addressed all of the comments.

kevinrushforth commented:
I addressed most of the feedback. There are still one or two more items I might update for this PR, but I'll capture the rest in follow-up issues.

nlisker commented on README-code-reviews.md:
```diff
@@ -37,46 +37,50 @@ In support of this, the following review policies are in effect. Many of these w
 * Make sure you understand why there was an issue to begin with, and why/how the proposed PR solves the issue
+* Carefully consider the risk of regression
+* Carefully consider any compatibility concerns
+* Check whether it adds any new public or protected API, even implicitly (such as a public method that overrides a protected method, or a class that is moved from a non-exported to an exported package); if it does, indicate that it needs a CSR
```
I think that if it looks like an oversight (the author forgot about the default constructor), it's better to indicate that than to indicate that the PR needs a CSR. Maybe something like:
"if it does and it doesn't look like an oversight, indicate that it needs a CSR"

kevinrushforth replied:
We have by now cleaned up our public API to avoid classes with an implicit no-arg constructor, so the only way this situation could arise in the future is if someone adds a new public class, which needs a CSR anyway.

I guess it could also happen if someone proposed to delete the no-arg constructor, but we don't allow deleting of non-terminally-deprecated API anyway. Maybe it's worth changing "adds any new" to "adds, removes, or modifies any"?

Somewhat related, these guidelines don't address what to look for when reviewing new API; perhaps a follow-on issue.

nlisker replied:
> Maybe it's worth changing "adds any new" to "adds, removes, or modifies any"?

This looks like a good idea.

kevinrushforth replied:
Done.

andy-goryachev-oracle reviewed (approved):
was going to comment on two spelling errors, but you fixed it.
looks good, thank you for moving this doc to the main repo as a markdown file!

nlisker commented on README-code-reviews.md:
```diff
@@ -37,46 +37,50 @@ In support of this, the following review policies are in effect. Many of these w
 * Check whether there is an automated test; if not, ask for one, if it is feasible
-* Make sure that the PR has executed the GHA tests and that they all pass; if they aren't being run, ask the PR author to enable GHA workflows
-* If the PR source branch hasn't synced up from master in a long time, or if there is an upstream commit not in the source branch that might interfere with the PR, ask the PR author to merge the latest upstream master.
+* Make sure that the PR has executed the GitHub Actions (GHA) tests; if they aren't being run, ask the PR author to enable GHA workflows; if the test fails on some platforms, check whether it is a real bug (sometimes a job fails becau se of GHA infrastucture changes or we see a spurious GHA failure)
```
Never mind, got simul-fixed.

kevinrushforth commented:
I filed the following follow-on bug to address and pending items:

[JDK-8332642](https://bugs.openjdk.org/browse/JDK-8332642): Additional improvements to contribution and code review guidelines

I think I have addressed all of the feedback so far, either by updating the docs or by adding an item to JDK-8332642.

kevinrushforth commented:
@johanvos Do you want to review this?

johanvos reviewed (approved):
Overall, this is a great improvement.
Reviewing code is a huge responsibility and it's good that we make it more clear now how we expect reviews to be executed.

kevinrushforth commented:
/integrate
